internal/adminapi/validators/consumer/credentials/validation.go

```diff
@@ -64,10 +101,41 @@ type Credential struct {
 // constraints on their respective types.
 type Index map[string]map[string]map[string]struct{}
 
-// Add will attempt to add a new Credential to the CredentialsTypeMap.
-// If that new credential is in violation of any constraints based on the
-// credentials already stored in the map, an error will be thrown.
-func (cs Index) Add(newCred Credential) error {
+// ValidateCredentialsForUniqueKeyConstraints will attempt to add a new Credential to the CredentialsTypeMap
+// and will validate it for both normal structure validation and for
+// unique key constraint violations.
+func (cs Index) ValidateCredentialsForUniqueKeyConstraints(consumerName string, secret *corev1.Secret) error {
+	// the indication of credential type is required to be present on all credentials.
+	credentialTypeB, ok := secret.Data[TypeKey]
+	if !ok {
+		return fmt.Errorf("missing required key %s", TypeKey)
+	}
+	credentialType := string(credentialTypeB)
+
+	// the additional key/values are optional, but must be validated
+	// for unique constraint violations. Using an index of credentials
+	// validation will be checked on any Add() to the index, so errors
+	// from this include the unique key constraint errors.
+	for k, v := range secret.Data {
+		if err := cs.add(Credential{
+			ConsumerName:      consumerName,
+			ConsumerNamespace: secret.Namespace,
+			Type:              credentialType,
+			Key:               k,
+			Value:             string(v),
+		}); err != nil {
+			return err
+		}
+	}
```

Comment:
Commentary only, no change request.

Creating a virtual credential with only a single field from the original and check that to see if that particular field conflicts rather than validating the entire credential seems odd, but I guess it works since the missing fields check is elsewhere.

IIRC the Kong DAO does support unique constraints on combined fields, i.e. you can have something like a uniqueness constraint on `username+password` rather than just `password` if you want, but I'm unsure and don't recall a credential that does that. Brief review suggests that's non-standard and requires a type-specific validation function. In either case, the current index system wouldn't support that anyway, so it's a moot point for now.

Reply:
I appreciate the thoughts on this. This was how the existing index worked prior to this PR and it has unit tests and everything wrapped up in it. I get that it's maybe a bit odd but without a strong reason to make a change I think we can just live with it for the moment since it's `internal/` anyway and we're not stuck with it.